docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
+	svcsdkapi "github.com/aws/aws-sdk-go/service/apigatewayv2/apigatewayv2iface"
++	"k8s.io/apimachinery/pkg/types"
++	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// +kubebuilder:rbac:groups=apigatewayv2.services.k8s.aws,resources=integrations,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apigatewayv2.services.k8s.aws,resources=integrations/status,verbs=get;update;patch
+
+var lateInitializeFieldNames = []string{}
+
++var (
++	_ = types.NamespacedName{}
++	_ = acksvcv1alpha1.Integration{}
++)
+
+...
+
+// DereferenceResource finds if there are any AWSResourceReference fields present
+// in the AWSResource passed in parameter and attempts to resolve those reference
+// fields into target field.
+// It returns a resolved AWSResource, boolean representing whether any reference
+// fields were present and an error if the passed AWSResource's reference fields
+// cannot be resolved
+func (rm *resourceManager) DereferenceResource(
+	ctx context.Context,
+	apiReader client.Reader,
+	res acktypes.AWSResource,
+) (acktypes.AWSResource, bool, error) {
+	ko := rm.concreteResource(res).ko.DeepCopy()
+	referencePresent := false
+	if ko.Spec.SecurityGroups != nil && ko.Spec.SecurityGroupIDs != nil {
+		return &resource{ko}, true, fmt.Errorf("'SecurityGroupIDs' field should not be present when using reference field 'SecurityGroups'")
+	}
+	// Checking Referenced Field SecurityGroups
+	if ko.Spec.SecurityGroups != nil && len(ko.Spec.SecurityGroups) > 0 {
+		referencePresent = true
+		resolvedReferences := []*string{}
+		for _, arr := range ko.Spec.SecurityGroups {
+			namespacedName := types.NamespacedName{Namespace: res.MetaObject().GetNamespace(), Name: *arr.Name}
+			obj := acksvcv1alpha1.API{}
+			err := apiReader.Get(ctx, namespacedName, &obj)
+			if err != nil {
+				return &resource{ko}, true, err
+			}
+			var refResourceSynced bool
+			for _, cond := range obj.Status.Conditions {
+				if cond.Type == ackv1alpha1.ConditionTypeResourceSynced && cond.Status == corev1.ConditionTrue {
+					refResourceSynced = true
+					break
+				}
+			}
+			if !refResourceSynced {
+				return &resource{ko}, true, fmt.Errorf("referenced 'API' resource " + *arr.Name + " does not have 'ACK.ResourceSynced' condition status 'True'")
+			}
+			if obj.Status.APIID == nil {
+				return &resource{ko}, true, fmt.Errorf("'Status.APIID' is not yet present for referenced 'API' resource " + *arr.Name)
+			}
+			resolvedReferences = append(resolvedReferences, obj.Status.APIID)
+		}
+		ko.Spec.SecurityGroupIDs = resolvedReferences
+	}
+	return &resource{ko}, referencePresent, nil
+}
+```
+
+## Alternate Approach
+* Change the type of member from `*string` to `*AWSResourceReference` and
+  `[]*string` to `[]*AWSResourceReference`
+* `AWSResourceReference` will either have the value for member field(string) or
+  contains the reference for another resource.
+* This approach was not chosen because it breaks the backward compatibility for
```

Comment:
could we maintain backward compat by enforcing the resource type and reference types are the same?

Reply:
Can you please elaborate? 
(The resource member type(*string/[]*string) cannot contain all information needed for referencing a resource.)

Reply:
Your other answers cleared this up for me -- misunderstanding on my end.